### PR TITLE
Set the cluster connection timeout to infinite by default

### LINKED ISSUE #306: [TRACKING ISSUE] Set cluster connect timeout to infinite by default
The tracking issue for the Java side PR.

**Set cluster connect timeout to infinite by default**

See https://github.com/hazelcast/hazelcast/pull/18094 for details.

---

After an internal discussion, we have decided to set the cluster
connect timeout to infinite by default. Also, the multiplier
for the sleep time between consecutive connect attempts is changed
to `1.05` from `1.0`. The new sleep times between connection
attempts can be seen in the following graph. https://www.desmos.com/calculator/hl7rnbk4gf
(x-axis: attempt number, y-axis: sleep time in seconds)

Also, updated some logging statements.

Note that, this is a breaking change for the users who rely on the client to stop
trying to connect cluster after some time, but we decided

### LINKED ISSUE #335: [TRACKING ISSUE] Increase default cluster timeout to 2 minutes
The tracking issue for the Java side PR.

**Increase default cluster timeout to 2 minutes**

See https://github.com/hazelcast/hazelcast/pull/18019 for details.

---

The old default tiemout was 20 seconds.
The change is done for environments like Kubernetes
where an Ip resolution/system restart could take
around a minute. We don't want a client to shutdown
in such cases, and we also want this to be out-of-the-box
behaviour.

Related config API

clientConfig.getConnectionStrategyConfig()
.getConnectionRetryConfig().getClusterConnectTimeoutMillis();

fixes https://github.com/hazelcast/hazelcast/issues/17904

### PR DESCRIPTION
**Pull request by mdumandag**
We decided to set the default cluster connection timeout to
infinite to have a better user experience out-of-the-box
for most of the users/use cases. Note that, this is a breaking
change for the users who rely on the client to shutdown after
some time. Affected users might set the `cluster_connect_timeout`
to a finite value.

Also, updated the default value of the `retry_multiplier` to back off
more after the client tries to connect to the target cluster for
some time.

Closes #335 
Closes #306

### CODE REVIEW

File: docs/client_connection_strategy.rst [hunk 1]
```diff
@@ -69,13 +69,14 @@ The following are configuration element descriptions:
 - ``retry_max_backoff``: Specifies the upper limit for the backoff in
   seconds. Its default value is ``30``. It must be non-negative.
 - ``retry_multiplier``: Factor to multiply the backoff after a failed
-  retry. Its default value is ``1``. It must be greater than or equal
+  retry. Its default value is ``1.05``. It must be greater than or equal
```
Comment (Kilo59) [inline docs/client_connection_strategy.rst:72]: What happened to the old [`connection_attempt_limit `](http://hazelcast.github.io/hazelcast-python-client/3.10/hazelcast.config.html#hazelcast.config.ClientNetworkConfig.connection_attempt_limit) ?

Is there no longer an option to control the number of connection attempts?

Comment (mdumandag) [reply docs/client_connection_strategy.rst:72]: Hi, this configuration option is removed from the client in the 4.0 release with a more enhanced exponential backoff feature. You can get the almost same behavior by configuring the `cluster_connect_timeout` parameter.

Comment (Kilo59) [reply docs/client_connection_strategy.rst:72]: In my use case, I just want to either attempt 1 or 2 connection attempts and then stop (as part of an integration test).

The exponential backup is a nice feature (and it will be useful in production), but it looks like the logic for limiting the connection attempts is now somewhat convoluted.

Comment (mdumandag) [reply docs/client_connection_strategy.rst:72]: I can agree on some of your points but let me clarify the `cluster_connect_timeout` a little bit. 

It is actually an absolute timeout for whole connection attempts.

Let say that you specified 3 possible member addresses in your configuration. Client first will store the time it started trying to connect the cluster. (Let say that it is `start_time`)

Client tries to connect one of the three. If it cannot connect any of them, then client tries to sleep a little bit. But before sleeping, it will check the the time passed since the first connection attempt(`time.time() - start_time`). If it is more than the `cluster_connect_timeout`, it will throw an exception saying that client could not connect to any of the members. If it is less than that, it will sleep and try to connect possible addresses again. This loops continues until the client reaches the `cluster_connect_timeout` or connects one of the members.

Comment (mdumandag) [reply docs/client_connection_strategy.rst:72]: The same logic applies to reconnection process

Comment (Kilo59) [reply docs/client_connection_strategy.rst:72]: @mdumandag thanks for the quick response and clarification, that's very helpful.

Comment (mdumandag) [reply docs/client_connection_strategy.rst:72]: No problem at all, happy to help you. Let me rephrase that part tomorrow

File: docs/client_connection_strategy.rst [hunk 2]
```diff
@@ -86,6 +87,7 @@ A pseudo-code is as follows:
     while (try_connect(connection_timeout)) != SUCCESS) {
         if (get_current_time() - begin_time >= CLUSTER_CONNECT_TIMEOUT) {
             // Give up to connecting to the current cluster and switch to another if exists.
+            // For the default values, CLUSTER_CONNECT_TIMEOUT is infinite.
```
Comment (yuce) [inline docs/client_connection_strategy.rst:90]: How about: "CLUSTER_CONNECT_TIMEOUT is infinite by default."?

File: hazelcast/connection.py [hunk 3]
```diff
@@ -280,11 +280,18 @@ def _trigger_cluster_reconnection(self):
             self._start_connect_to_cluster_thread()
 
     def _init_wait_strategy(self, config):
+        cluster_connect_timeout = config.cluster_connect_timeout
+        if cluster_connect_timeout == -1:
+            # If the no timeout is specified by the
+            # user, or set to -1 explicitly, set
+            # the timeout to infinite.
+            cluster_connect_timeout = six.MAXSIZE
```
Comment (yuce) [inline hazelcast/connection.py:288]: As far as I've checked `sys.maxsize` is available on all versions of CPython. On Python 2.7 and Python 3.6 both evaluate `sys,maxsize` to `9223372036854775807`. So, I think it would be better to replace `sys.MAXSIZE` with `sys.maxsize`.

File: hazelcast/config.py [hunk 4]
```diff
@@ -531,8 +531,8 @@ def __init__(self):
         self._retry_initial_backoff = 1.0
         self._retry_max_backoff = 30.0
         self._retry_jitter = 0.0
-        self._retry_multiplier = 1.0
-        self._cluster_connect_timeout = 120.0
+        self._retry_multiplier = 1.05
```
Comment (yuce) [inline hazelcast/config.py:534]: Can we move `1.05` to a global variable, something like `DEFAULT_RETRY_MULTIPLIER` and replace instances of it?

File: hazelcast/config.py [hunk 5]
```diff
@@ -531,8 +531,8 @@ def __init__(self):
...
+        self._cluster_connect_timeout = -1
```
Comment (yuce) [inline hazelcast/config.py:535]: Can we move `-1` to a global variable, something like `DEFAULT_CLUSTER_CONNECT_TIMEOUT` and replace instances of it?

File: docs/client_connection_strategy.rst [hunk 6]
```diff
@@ -69,13 +69,14 @@ The following are configuration element descriptions:
...
-  retry. Its default value is ``1``. It must be greater than or equal
+  retry. Its default value is ``1.05``. It must be greater than or equal
   to ``1``.
 - ``retry_jitter``: Specifies by how much to randomize backoffs. Its
   default value is ``0``. It must be in range ``0`` to ``1``.
 - ``cluster_connect_timeout``: Timeout value in seconds for the client
-  to give up to connect to the current cluster. Its default value is
-  ``120``.
+  to give up connecting to the cluster. Its default value is
+  ``-1``. For the default value, client will not stop trying to connect
```
Comment (yuce) [inline docs/client_connection_strategy.rst:78]: I think we can rephrase "For the default value ..." as "By default ..."

**Review by yuce — CHANGES_REQUESTED** [review]
The PR looks good other than a few nitpicks.

File: docs/client_connection_strategy.rst (hunk 6 above)
Comment (yuce) [reply docs/client_connection_strategy.rst:78]: Maybe even better: "The client will continuously try to connect by default."

**Review by yuce — APPROVED** [review]
LGTM

**Comment by mdumandag** [comment]
Thanks for the review @yuce
